packages/cdktf/lib/testing/matchers.ts

```diff
@@ -0,0 +1,179 @@
+import * as fs from "fs";
+import * as path from "path";
+import { execSync } from "child_process";
+
+const terraformBinaryName = process.env.TERRAFORM_BINARY_NAME || "terraform";
+
+export interface TerraformConstructor {
+  new (...args: any[]): any;
+  terraformResourceType: string;
+}
+export type SynthesizedStack = {
+  resource: Record<string, any>;
+  data: Record<string, any>;
+};
+export type MatcherReturn = { message: () => string; pass: boolean };
+
+function assertElementWithProperties(
+  type: keyof SynthesizedStack,
+  received: string,
+  itemType: TerraformConstructor,
+  properties: Record<string, any> = {}
+) {
+  const items = Object.entries(
+    (JSON.parse(received) as SynthesizedStack)[type] || {}
+  );
+
+  const pass = items.some(([type, values]) => {
+    if (type !== itemType.terraformResourceType) {
+      return false;
+    }
+
+    return Object.values(values).some((item: any) => {
+      return Object.entries(properties).every(([key, value]) => {
+        // TODO: snake case the key
+        return item[key] === value;
+      });
+    });
+  });
+
+  if (pass) {
+    return {
+      pass,
+      message: () =>
+        `Expected ${
+          itemType.terraformResourceType
+        } not to be present in synthesised stack with properties ${JSON.stringify(
+          properties
+        )}`,
+    };
+  } else {
+    return {
+      message: () =>
+        `Expected ${
+          itemType.terraformResourceType
+        } to be present in synthesised stack with properties ${JSON.stringify(
+          properties
+        )}`,
+      pass,
+    };
+  }
+}
+
+export function toHaveDataSourceWithProperties(
+  received: string,
+  resourceType: TerraformConstructor,
+  properties: Record<string, any> = {}
+): MatcherReturn {
+  return assertElementWithProperties(
+    "data",
+    received,
+    resourceType,
+    properties
+  );
+}
+
+export function toHaveResourceWithProperties(
+  received: string,
+  resourceType: TerraformConstructor,
+  properties: Record<string, any> = {}
+): MatcherReturn {
+  return assertElementWithProperties(
+    "resource",
+    received,
+    resourceType,
+    properties
+  );
+}
+
+export function toBeValidTerraform(received: string): MatcherReturn {
```

Comment:
as this matcher seems to work on a directory, I'd be explicit here (if possible) and use `toBeValidTerraformDirectory` or likewise (I'm not quite happy with that alternative name yet, but something like that).
Or can we use types to narrow that down?

Reply:
I think it's unclear that fullSynth returns a directory, if we want to rename the matcher, we should also find a better name for fullSynth 🤔  Open for suggestions